From 9c5d129372c37cf585c2113bc4272fa6a91a3789 Mon Sep 17 00:00:00 2001
From: Ninad Naik <quic_ninanaik@quicinc.com>
Date: Tue, 3 Dec 2024 18:55:55 +0530
Subject: [PATCH] PENDING: firmware: qcom-scm: Introduce new locking mechanism
 for SCM driver

With the existing global mutex lock, there is a possibility that it
could resulty into a deadlock when SMCInvoke and smc call go to
firmware in the same order and both wait on same WAITQ in firmware.

Change-Id: Id4bf99867682f630573e74e5b2556e32583dbc71
Signed-off-by: Venkatakrishnaiah Pari <quic_vpari@quicinc.com>
Signed-off-by: Jian Shu <quic_jianshu@quicinc.com>
Signed-off-by: Ninad Naik <quic_ninanaik@quicinc.com>
---
 drivers/firmware/qcom_scm-legacy.c |  5 ++---
 drivers/firmware/qcom_scm-smc.c    | 11 ++---------
 drivers/firmware/qcom_scm.c        |  6 ++++++
 drivers/firmware/qcom_scm.h        |  3 +++
 4 files changed, 13 insertions(+), 12 deletions(-)

diff --git a/drivers/firmware/qcom_scm-legacy.c b/drivers/firmware/qcom_scm-legacy.c
index 029e6d117cb8..1cea9547ca12 100644
--- a/drivers/firmware/qcom_scm-legacy.c
+++ b/drivers/firmware/qcom_scm-legacy.c
@@ -15,7 +15,6 @@
 
 #include "qcom_scm.h"
 
-static DEFINE_MUTEX(qcom_scm_lock);
 
 
 /**
@@ -173,11 +172,11 @@ int scm_legacy_call(struct device *dev, const struct qcom_scm_desc *desc,
 	smc.args[1] = (unsigned long)&context_id;
 	smc.args[2] = cmd_phys;
 
-	mutex_lock(&qcom_scm_lock);
+	down(&qcom_scm_sem_lock);
 	__scm_legacy_do(&smc, &smc_res);
 	if (smc_res.a0)
 		ret = qcom_scm_remap_error(smc_res.a0);
-	mutex_unlock(&qcom_scm_lock);
+	up(&qcom_scm_sem_lock);
 	if (ret)
 		goto out;
 
diff --git a/drivers/firmware/qcom_scm-smc.c b/drivers/firmware/qcom_scm-smc.c
index f4fa8b74d1f3..78bcb7eb491a 100644
--- a/drivers/firmware/qcom_scm-smc.c
+++ b/drivers/firmware/qcom_scm-smc.c
@@ -25,8 +25,6 @@ struct arm_smccc_args {
 	unsigned long args[8];
 };
 
-static DEFINE_MUTEX(qcom_scm_lock);
-
 #define QCOM_SCM_EBUSY_WAIT_MS 30
 #define QCOM_SCM_EBUSY_MAX_RETRY 20
 
@@ -126,16 +124,11 @@ static int __scm_smc_do_quirk_handle_waitq(struct device *dev, struct arm_smccc_
 			}
 
 			if (res->a0 == QCOM_SCM_WAITQ_SLEEP) {
-				if (skip_mutex)
-					mutex_unlock(&qcom_scm_lock);
 
 				ret = qcom_scm_wait_for_wq_completion(dev_get_drvdata(dev), wq_ctx);
 				if (ret)
 					return ret;
 
-				if (skip_mutex)
-					mutex_lock(&qcom_scm_lock);
-
 				fill_wq_resume_args(&fill, smc_call_ctx);
 				smc = &fill;
 			} else if (res->a0 == QCOM_SCM_WAITQ_WAKE) {
@@ -161,11 +154,11 @@ static int __scm_smc_do(struct device *dev, struct arm_smccc_args *smc,
 	}
 
 	do {
-		mutex_lock(&qcom_scm_lock);
+		down(&qcom_scm_sem_lock);
 
 		ret = __scm_smc_do_quirk_handle_waitq(dev, smc, res, skip_mutex);
 
-		mutex_unlock(&qcom_scm_lock);
+		up(&qcom_scm_sem_lock);
 
 		if (ret)
 			return ret;
diff --git a/drivers/firmware/qcom_scm.c b/drivers/firmware/qcom_scm.c
index 6480f0f0ea4b..1d394764b805 100644
--- a/drivers/firmware/qcom_scm.c
+++ b/drivers/firmware/qcom_scm.c
@@ -48,11 +48,14 @@ struct qcom_scm {
 	struct mutex scm_bw_lock;
 	int scm_vote_count;
 
+	u64 call_ctx_cnt;
 	u64 dload_mode_addr;
 
 	struct qcom_tzmem_pool *mempool;
 };
 
+DEFINE_SEMAPHORE(qcom_scm_sem_lock, 1);
+
 #define SHMBRIDGE_RESULT_NOTSUPP		4
 
 /* Each bit configures cold/warm boot address for one of the 4 CPUs */
@@ -1715,6 +1718,8 @@ static void __check_fw_for_skip_mutex_support(void)
 
 	num_simultaneous_requests = res.result[0] & 0xFF;
 
+	__scm->call_ctx_cnt = res.result[0] & 0xFF;
+
 	__scm->fw_supports_skip_mutex = (num_simultaneous_requests >= MIN_SIMULTANEOUS_REQS);
 }
 
@@ -1893,6 +1898,7 @@ static int qcom_scm_probe(struct platform_device *pdev)
 
 	__get_convention();
 	__check_fw_for_skip_mutex_support();
+	sema_init(&qcom_scm_sem_lock, (int)__scm->call_ctx_cnt);
 
 	/*
 	 * If requested enable "download mode", from this point on warmboot
diff --git a/drivers/firmware/qcom_scm.h b/drivers/firmware/qcom_scm.h
index f06381eece5f..7811528ab418 100644
--- a/drivers/firmware/qcom_scm.h
+++ b/drivers/firmware/qcom_scm.h
@@ -5,6 +5,8 @@
 #ifndef __QCOM_SCM_INT_H
 #define __QCOM_SCM_INT_H
 
+#include <linux/semaphore.h>
+
 struct qcom_tzmem_pool;
 
 enum qcom_scm_convention {
@@ -15,6 +17,7 @@ enum qcom_scm_convention {
 };
 
 extern enum qcom_scm_convention qcom_scm_convention;
+extern struct semaphore qcom_scm_sem_lock;
 
 #define MAX_QCOM_SCM_ARGS 10
 #define MAX_QCOM_SCM_RETS 3
-- 
2.17.1

